File: Python FastApi/routers/feature_options.py
from fastapi import APIRouter, Depends, HTTPException, status
from motor.motor_asyncio import AsyncIOMotorDatabase
from bson import ObjectId
from typing import List, Dict
from datetime import datetime

from dependencies import get_db, get_current_user, require_role, Role, public
from models.feature_option import FeatureOptionCreate, FeatureOptionUpdate, FeatureOptionResponse

router = APIRouter()

@router.post("/", response_model=FeatureOptionResponse)
async def create_feature_option(
    feature_data: FeatureOptionCreate,
    db: AsyncIOMotorDatabase = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    # Check for duplicate
    existing_option = await db.feature_options.find_one({
        "category": feature_data.category,
        "value": feature_data.value
    })
    if existing_option:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Bu özellik zaten mevcut."
        )
    
    feature_doc = {
        **feature_data.dict(),
        "created_at": datetime.utcnow(),
        "updated_at": datetime.utcnow()
    }
    
    result = await db.feature_options.insert_one(feature_doc)
    feature_doc["_id"] = result.inserted_id
    
    return FeatureOptionResponse(**feature_doc)

@router.get("/", response_model=Dict[str, List[str]])
@public()
async def get_all_feature_options_grouped(db: AsyncIOMotorDatabase = Depends(get_db)):
    options = await db.feature_options.find().to_list(None)
    
    grouped_options = {}
    for option in options:
        category = option["category"]
        value = option["value"]
        if category not in grouped_options:
            grouped_options[category] = []
        grouped_options[category].append(value)
    
    return grouped_options

@router.get("/findall", response_model=List[FeatureOptionResponse])
async def get_all_feature_options(
    db: AsyncIOMotorDatabase = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    options = await db.feature_options.find().to_list(None)
    return [FeatureOptionResponse(**option) for option in options]

@router.get("/{id}", response_model=FeatureOptionResponse)
async def get_feature_option(
    id: str,
    db: AsyncIOMotorDatabase = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    option = await db.feature_options.find_one({"_id": ObjectId(id)})
    if not option:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Özellik bulunamadı"
        )
    return FeatureOptionResponse(**option)

@router.put("/{id}", response_model=FeatureOptionResponse)
async def update_feature_option(
    id: str,
    feature_data: FeatureOptionUpdate,
    db: AsyncIOMotorDatabase = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    option = await db.feature_options.find_one({"_id": ObjectId(id)})
    if not option:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Özellik bulunamadı"
        )
    
    # Check for potential duplicate
    update_category = feature_data.category or option["category"]
    update_value = feature_data.value or option["value"]
    
    if feature_data.category or feature_data.value:
        potential_duplicate = await db.feature_options.find_one({
            "category": update_category,
            "value": update_value,
            "_id": {"$ne": ObjectId(id)}
        })
        if potential_duplicate:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Güncelleme sonucunda başka bir özellikle aynı olacak. Lütfen farklı bir değer girin."
            )
    
    update_data = {k: v for k, v in feature_data.dict(exclude_unset=True).items() if v is not None}
    update_data["updated_at"] = datetime.utcnow()
    
    if update_data:
        await db.feature_options.update_one(
            {"_id": ObjectId(id)},
            {"$set": update_data}
        )
    
    updated_option = await db.feature_options.find_one({"_id": ObjectId(id)})
    return FeatureOptionResponse(**updated_option)

@router.delete("/{id}")
async def delete_feature_option(
    id: str,
    db: AsyncIOMotorDatabase = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    result = await db.feature_options.delete_one({"_id": ObjectId(id)})
    if result.deleted_count == 0:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Özellik bulunamadı"
        )
    
    return {"message": "Özellik başarı ile silindi"}

File: nodejs/routes/feature-options.js
const express = require('express');
const FeatureOption = require('../models/FeatureOption');
const { authorizeRoles } = require('../middleware/auth');

const router = express.Router();

/**
 * @swagger
 * /api/feature-options:
 *   post:
 *     summary: Create a new feature option
 *     tags: [Feature Options]
 *     security:
 *       - cookieAuth: []
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required:
 *               - category
 *               - value
 *             properties:
 *               category:
 *                 type: string
 *               value:
 *                 type: string
 *     responses:
 *       201:
 *         description: Feature option created successfully
 *       409:
 *         description: Feature option already exists
 */
router.post('/', authorizeRoles('admin', 'member'), async (req, res, next) => {
  try {
    const { category, value } = req.body;

    const newOption = new FeatureOption({ category, value });
    await newOption.save();

    res.status(201).json(newOption);
  } catch (error) {
    if (error.code === 11000) {
      return res.status(409).json({ message: 'Bu özellik zaten mevcut.' });
    }
    next(error);
  }
});

/**
 * @swagger
 * /api/feature-options:
 *   get:
 *     summary: Get all feature options grouped by category
 *     tags: [Feature Options]
 *     responses:
 *       200:
 *         description: Grouped feature options
 */
router.get('/', async (req, res, next) => {
  try {
    const allOptions = await FeatureOption.find();
    
    const groupedOptions = allOptions.reduce((acc, option) => {
      const { category, value } = option;
      if (!acc[category]) acc[category] = [];
      acc[category].push(value);
      return acc;
    }, {});

    res.json(groupedOptions);
  } catch (error) {
    next(error);
  }
});

/**
 * @swagger
 * /api/feature-options/findall:
 *   get:
 *     summary: Get all feature options (ungrouped)
 *     tags: [Feature Options]
 *     security:
 *       - cookieAuth: []
 *     responses:
 *       200:
 *         description: List of all feature options
 */
router.get('/findall', authorizeRoles('admin', 'member'), async (req, res, next) => {
  try {
    const options = await FeatureOption.find();
    res.json(options);
  } catch (error) {
    next(error);
  }
});

/**
 * @swagger
 * /api/feature-options/{id}:
 *   get:
 *     summary: Get feature option by ID
 *     tags: [Feature Options]
 *     security:
 *       - cookieAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: Feature option details
 *       404:
 *         description: Feature option not found
 */
router.get('/:id', authorizeRoles('admin', 'member'), async (req, res, next) => {
  try {
    const { id } = req.params;
    
    const option = await FeatureOption.findById(id);
    if (!option) {
      return res.status(404).json({ message: 'Özellik bulunamadı' });
    }

    res.json(option);
  } catch (error) {
    next(error);
  }
});

/**
 * @swagger
 * /api/feature-options/{id}:
 *   put:
 *     summary: Update feature option
 *     tags: [Feature Options]
 *     security:
 *       - cookieAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             properties:
 *               category:
 *                 type: string
 *               value:
 *                 type: string
 *     responses:
 *       200:
 *         description: Feature option updated successfully
 *       404:
 *         description: Feature option not found
 *       409:
 *         description: Duplicate feature option
 */
router.put('/:id', authorizeRoles('admin', 'member'), async (req, res, next) => {
  try {
    const { id } = req.params;
    const updateDto = req.body;

    const existingOption = await FeatureOption.findById(id);
    if (!existingOption) {
      return res.status(404).json({ message: 'Özellik bulunamadı' });
    }

    // Check for duplicates
    if (updateDto.category || updateDto.value) {
      const potentialDuplicate = await FeatureOption.findOne({
        category: updateDto.category || existingOption.category,
        value: updateDto.value || existingOption.value,
      });

      if (potentialDuplicate && potentialDuplicate._id.toString() !== id) {
        return res.status(409).json({ 
          message: 'Güncelleme sonucunda başka bir özellikle aynı olacak. Lütfen farklı bir değer girin.' 
        });
      }
    }

    // Update option
    Object.assign(existingOption, updateDto);
    await existingOption.save();

    res.json(existingOption);
  } catch (error) {
    next(error);
  }
});

/**
 * @swagger
 * /api/feature-options/{id}:
 *   delete:
 *     summary: Delete feature option
 *     tags: [Feature Options]
 *     security:
 *       - cookieAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: Feature option deleted successfully
 *       404:
 *         description: Feature option not found
 */
router.delete('/:id', authorizeRoles('admin', 'member'), async (req, res, next) => {
  try {
    const { id } = req.params;
    
    const result = await FeatureOption.findByIdAndDelete(id);
    if (!result) {
      return res.status(404).json({ message: 'Özellik bulunamadı' });
    }

    res.json({ message: 'özellik başarı ile silindi' });
  } catch (error) {
    next(error);
  }
});

module.exports = router;